specs/opcode/1dSAR.md

```diff
@@ -0,0 +1,150 @@
+
+More formally, the variables are defined as follows:
+
+```
+is_neg = is_neg(a)
+shf0 = bytes_to_fq(shift.le_bytes[:1])
+shf_div64 = shift // 64
+shf_mod64 = shift % 64
+p_lo = 1 << shf_mod64
+p_hi = 1 << (64 - shf_mod64)
+# The new bits are set to 1 if negative.
+p_top = is_neg * (0xFFFFFFFFFFFFFFFF - p_hi + 1))
+shf_lt256 = is_zero(sum(shift[1:]))
+a64s = word_to_64s(a)
+a64s_lo[idx] = a64s[idx] % p_lo
+a64s_hi[idx] = a64s[idx] / p_lo
+```
+
+`b64s` could be calculated by these variables:
+
+* Initialized `b64s`: It should be `[0xFFFFFFFFFFFFFFFF] * 4` if `is_neg`, `[0] * 4` otherwise.
+* `b64s[3 - shf_div64]`: It could be calculated by `a64s_hi[3] + p_top`.
+* `b64s[k]`: It could be calculated by `a64s_hi[k + shf_div64] + a64s_lo[k + shf_div64 + 1] * p_hi` where `k < 3 - shf_div64`.
+
+Now putting things together, the constraints can be constructed as follows:
+
+1. `a64s` and `b64s` constraints:
+
+* First calculate `shf_lt256` as `is_zero(sum(shift[1:]))`.
+* `a64s[idx]`: It should be equal to `from_bytes(a[8 * idx : 8 * (idx + 1)])` where idx in `(0, 1, 2, 3)`.
+* `b64s[idx] * shf_lt256 + is_neg * (1 - shf_lt256) * 0xFFFFFFFFFFFFFFFF`: It should be equal to `from_bytes(b[8 * idx : 8 * (idx + 1)])` where idx in `(0, 1, 2, 3)`.
+
+2. `a64s_lo` and `a64s_hi` constraints:
+
+* `a64s[idx]`: It should be equal to `a64s_lo[idx] + a64s_hi[idx] * p_lo`.
+* `a64s_lo[idx]`: It should always be less than `p_lo` (`a64s_lo[idx] < p_lo`).
+
+3. Merge constraints:
+
+* First create four `IsZero` gadgets:
+```
+shf_div64_eq0 = is_zero(shf_div64)
+shf_div64_eq1 = is_zero(shf_div64 - 1)
+shf_div64_eq2 = is_zero(shf_div64 - 2)
+shf_div64_eq3 = is_zero(shf_div64 - 3)
+```
+
+* `b64s[0]` should be equal to:
+```
+(a64s_hi[0] + a64s_lo[1] * p_hi) * shf_div64_eq0 +
+  (a64s_hi[1] + a64s_lo[2] * p_hi) * shf_div64_eq1 +
+  (a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq2 +
+  (a64s_hi[3] + p_top) * shf_div64_eq3 +
+  is_neg * 0xFFFFFFFFFFFFFFFF * (1 - shf_div64_eq0 - shf_div64_eq1 - shf_div64_eq2 - shf_div64_eq3)
+```
+
+* `b64s[1]` should be equal to:
+```
+(a64s_hi[1] + a64s_lo[2] * p_hi) * shf_div64_eq0 +
+  (a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq1 +
+  (a64s_hi[3] + p_top) * shf_div64_eq2 +
+  is_neg * 0xFFFFFFFFFFFFFFFF * (1 - shf_div64_eq0 - shf_div64_eq1 - shf_div64_eq2)
+```
+
+* `b64s[2]` should be equal to:
+```
+(a64s_hi[2] + a64s_lo[3] * p_hi) * shf_div64_eq0 +
+  (a64s_hi[3] + p_top) * shf_div64_eq1 +
+  is_neg * 0xFFFFFFFFFFFFFFFF * (1 - shf_div64_eq0 - shf_div64_eq1)
+```
+
+* `b64s[3]` should be equal to:
+```
+(a64s_hi[3] + p_top) * shf_div64_eq0 +
+  is_neg * 0xFFFFFFFFFFFFFFFF * (1 - shf_div64_eq0)
+```
+
+4. `shift[0]` constraint:
+
+* `shift[0]`: It should be equal to `shf_mod64 + shf_div64 * 64`.
+
+5. `Pow2` table look up:
+
+* First build `Pow2` table by tuple `(value, value_pow)` which meets `value_pow == pow(2, value)`.
+
+* Look up for `(shf_mod64, p_lo)` and `(64 - shf_mod64, p_hi)`.
+
```

Comment:
Added `is_neg` constraints description to Markdown doc (and corresponding code as below):
```
. `is_neg` constraints:

* Constrain `is_neg` must be a boolean (0 or 1).
* Assign `sign_byte = 255` if `is_neg == 1`, otherwise `sign_byte = 0`, then lookup `SignByte` table by tuple `(a_highest_byte, sign_byte)`.
```